lesson12.md

```diff
@@ -0,0 +1,412 @@
+
+It is possible, but it's a mess.
+<!-- .element: class="fragment"  -->
+
+---
+
+### Objects
+
+An **object**, in Javascript, is a group (or a list) of *key* / *value* pairs
+
+![Pears](images/pears.jpg) <!-- .element width="300" style="display: block; margin: 0 auto; width: 300px;" -->
+
+
+---
+
+### Objects
+
+Example:
+
+```js
+let person = {
+  firstName: 'Carlo',
+  lastName: 'Trimarchi',
+  age: 39
+};
+```
+
+Can you guess how we can create an object?
+
+---
+
+### Objects
+
+To create an object we use curly braces `{}`.
+```js
+let emptyObject = {};
+```
+
+We put any key/value pair inside the `{}`
+
+```js
+let person = {
+  name: 'Carlo'
+  // the key is 'name', the value is 'Carlo'
+};
+```
+
+---
+
+### Objects
+
+We can think of key/value pairs as words in a dictionary
+
+![Pears](images/dictionary.jpg) <!-- .element width="500" style="display: block; margin: 0 auto;" -->
+
+* a word is the **key**
+* the definition is the **value**
+
+---
+
+### Objects
+
+key/value pairs are also referred to as **properties**
+
+```js
+let person = {
+  firstName: 'Carlo', // comma here
+  age: 39, // also here
+  isVegetarian: true // this is the last one, no comma
+};
+```
+* can you list all the properties in this **person** object?
+* and its keys?
+* and its values?
+
+We can add as many properties as we want.
+<!-- .element: class="fragment"  -->
+
+---
+
+### Objects: examples
+
+```js
+
+let book = {
+  title: 'The Lord of the Rings',
+  author: 'J. R. R. Tolkien',
+  publicationYear: 1954,
+  pages: 1216
+};
+
+```
+
+Can you think of other things we can represent with objects?
+<!-- .element: class="fragment"  -->
+
+---
+
+### Objects: examples
+
+* movies
+<!-- .element: class="fragment"  -->
+* passport / identity card
+<!-- .element: class="fragment"  -->
+* cooking recipe
+<!-- .element: class="fragment"  -->
+* product in an e-commerce
+<!-- .element: class="fragment"  -->
+* much more...
+<!-- .element: class="fragment"  -->
+
+---
+
+### Objects: exercise
+
+Try it yourself:
+
+create an object called "`me`" containing the following information about you:
+* name
+* age
+* hair color
+* eye color
+
+Log your object to the console
+
+---
+
+### Objects vs arrays
+
+```js
+let meArray = [39, "Carlo", "Trimarchi", "Italian", "Messina",
+          "green", "black", false, true,
+          ['reading', 'music', 'guitar'],
+          ['pizza', 'pasta', 'lasagna']];
+
+let me = {
+  age: 39,
+  firstName:  "Carlo",
+  lastName:  "Trimarchi",
+  nationality:  "Italian",
+  cityOfBirth:  "Messina",
+  favouriteColour:  "green",
+  hairColour:  "black",
+  canSpeakGerman:  false,
+  isVegetarian:  true,
+  hobbies:  ['reading', 'music', 'guitar'],
+  favouriteFoods:  ['pizza', 'pasta', 'lasagna']
+};
+```
+
+Objects let us give **structure** and **meaning** to our data
+
+---
+
+### Objects: functions
+
+We can also add functions as properties!
+
+```js
+let dog = {
+  name: "Bob",
+  breed: "beagle",
+  age: 3,
+  speak: function() {
```

Comment:
Just a thought from previous semesters, I think functions as properties of objects is usually one of the more confusing thing for students. It could be a lecture all on its own including practice exercises. I think it's because functions in general are apparently harder to grasp.

Reply:
@owenlow you're right. My idea is to just mention them and introducing the possibility of having functions inside objects. Then, I'm sure we should do at least another lesson on objects and we would have time to expand on the concept again

Reply:
In addition this is an anonymous function value. Something they haven't seen before.

Reply:
so, shall I just remove the function part?

Reply:
No, definitely keep because it's key to understand DOM. The problem is that the concept of functions as variables is very confusing so we never managed to explain it right :/ In this case, mention it as special syntax for functions? E.g. this is how you associate a function with a key-value pair? More to come if people as questions?

Reply:
yes, I went through the slides of previous semesters and we never explain anonymous functions properly. I'll just introduce the syntax whithout going too much into theory and maybe show a few examples of how it's used

Reply:
I never realized, but I don't think we've ever gone over anonymous functions explicitly as a topic. Maybe the class after this one could be specifically about anonymous functions, and how they are used in objects.

Then afterwards we can start the DOM stuff and explain how document is a global object that has functions, just like the exercises we've done in class.

Reply:
we could introduce the higher order functions on arrays and first how you can put named functions in there and then transition to anonymous and big-arrow functions?
And then assigning functions to variables/object-values should be easy-peasy.

What kind of experiences have you made in the last semesters with this topic?
Did student get it or was it too far of a leap at this stage?
"What? I can pass functions to functions :exploding_head: "

Reply:
We never taught higher order functions and neither big-arrow functions. They will learn them in the React course.
We can teach a lot of this stuff, but if then they don't have a way to practice it, there's no point because it's forgotten right away.
Functions inside objects seems to me a reasonable step forward from normal functions. At this level, they don't need to know too much of the theory behind, just the syntax and some examples to see them in use. The theory and technicalities behind them can come at a later stage